packages/graphiql/src/state/sessionActions.ts

```diff
@@ -0,0 +1,104 @@
+import { EditorContexts } from './types';
+
+export enum SessionActionTypes {
+  OperationRequested = 'OperationRequested',
+  EditorLoaded = 'EditorLoaded',
+  OperationChanged = 'OperationChanged',
+  VariablesChanged = 'VariablesChanged',
+  OperationSucceeded = 'OperationSucceeded',
+  OperationErrored = 'OperationErrored',
+}
+
+export type ErrorPayload = { error: Error; sessionId: number };
+export type SuccessPayload = { sessionId: number; result: string };
+export type ChangeValuePayload = { sessionId: number; value: string };
+
+export type SessionAction =
+  | OperationRequestedAction
+  | EditorLoadedAction
+  | OperationChangedAction
+  | VariablesChangedAction
+  | OperationSucceededAction
+  | OperationErroredAction;
+
+export type OperationRequestedAction = {
+  type: SessionActionTypes.OperationRequested;
+};
+
+export const operationRequestAction = (): OperationRequestedAction => ({
+  type: SessionActionTypes.OperationRequested,
+});
+
+export type EditorLoadedAction = {
+  type: SessionActionTypes.EditorLoaded;
+  payload: {
+    context: EditorContexts;
+    editor: CodeMirror.Editor;
+  };
+};
```

Comment:
re: https://github.com/graphql/graphiql/pull/1462#discussion_r403643878

how 'bout something like this?
```suggestion
export type EditorLoadedAction = ReturnType<typeof editorLoadedAction>;
```

Reply:
@cshaver yeah i do think i like that better